### PR TITLE
Refactor: migrate interval function for datavalue2

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

## Changelog

- Improvement

## Related Issues

Fixes #issue

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/functions/src/scalars/expressions/binary.rs [hunk 1]
```diff
@@ -15,8 +15,122 @@
 use std::marker::PhantomData;
 
 use common_datavalues2::prelude::*;
+use common_exception::ErrorCode;
 use common_exception::Result;
 
+#[derive(Debug, Clone)]
+pub struct EvalContext {
```
Comment (sundy-li) [inline common/functions/src/scalars/expressions/binary.rs:22]: Let's move it into a new file, cause unary may also need it.

File: common/functions/src/scalars/dates/macros.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+#[macro_export]
+macro_rules! define_date_add_year_months {
+    ($l: ident, $r: ident, $ctx: ident, $date_type:ident, $op: expr) => {{
+        let factor = $ctx.factor;
+        let epoch = NaiveDate::from_ymd(1970, 1, 1);
+        let naive = epoch.checked_add_signed(Duration::days($l as i64));
+        if naive.is_none() {
+            $ctx.error = Some(ErrorCode::Overflow(format!(
```
Comment (sundy-li) [inline common/functions/src/scalars/dates/macros.rs:67]: Better to add a `set_error` function:

```
fn set_error(&mut self, e: ErrorCode) {
    if self.error.is_none() {
       self.error = e;
    }
}
```

Comment (sundy-li) [reply common/functions/src/scalars/dates/macros.rs:67]: This keeps the first error.

File: common/functions/src/scalars/expressions/binary.rs [hunk 3]
```diff
@@ -17,6 +17,101 @@ use std::marker::PhantomData;
 use common_datavalues2::prelude::*;
 use common_exception::Result;
 
+use super::EvalContext;
+
+pub trait ScalarBinaryFunction2<L: Scalar, R: Scalar, O: Scalar> {
```
Comment (sundy-li) [inline common/functions/src/scalars/expressions/binary.rs:22]: I think we just introduce `ctx` into `ScalarBinaryFunction` and refactor others using this function is better.

`ScalarBinaryExpression2` & `ScalarBinaryFunction2` can be merged into `ScalarBinaryExpression` & `ScalarBinaryRefFunction`

Comment (zhyass) [reply common/functions/src/scalars/expressions/binary.rs:22]: > I think we just introduce `ctx` into `ScalarBinaryFunction` and refactor others using this function is better.
> 
> `ScalarBinaryExpression2` & `ScalarBinaryFunction2` can be merged into `ScalarBinaryExpression` & `ScalarBinaryRefFunction`

Yes, I will merged after add unit and stateless tests

**Review by junli1026 — APPROVED** [review]

**Review by sundy-li — APPROVED** [review]

**Pull request merged**